### ISSUE TITLE
Improve / cache characteristic polynomial computation

### ISSUE THREAD

**Issue by sritchie** [sicmutils/sicmutils#209]
A note from @littleredcomputer on a previous PR:

> Before I forget: I was looking at this and thinking, "it would be cool if `(characteristic-polynomial M)` returned a function of `x` that was polynomial in `x`, without having to name it. Of course, that is easily done, by just having the function return (fn [x] ...), but then it does a (potentially vast) amount of computation every time you call it. If only there were a way to record that computation in a way that could be played back on `'x` or `3`. I think that is something we should think about; it would have application to our numerics and to differentiation itself.

One way to do this that would work today would be to pass a symbol, then simplify and compile the resulting symbolic expression:

```clojure
(defn characteristic-fn [M]
  (compile/compile-univariate-fn
   (partial m/characteristic-polynomial M)))
```

This would NOT, unlike your suggestion, be able to handle `'x` or `3`, just `3`, since we compile the native math functions in.

But we COULD make an alternative `sci-context` for the `compile` namespace that would stick the generics back in, instead of the real versions. @littleredcomputer would this be a good thing?

Before I thought of this approach, I had found a presentation called ["Efficient Computation of the Characteristic Polynomial"](https://ljk.imag.fr/membres/Clement.Pernet/Publications/charpoly_issac05_slides.pdf), so I'll include that here for reference.

**Comment by sritchie** [comment]
"In mathematics (linear algebra), the Faddeev–LeVerrier algorithm is a recursive method to calculate the coefficients of the characteristic polynomial {\displaystyle p(\lambda )=\det(\lambda I_{n}-A)}{\displaystyle p(\lambda )=\det(\lambda I_{n}-A)} of a square matrix, A, named after Dmitry Konstantinovich Faddeev and Urbain Le Verrier. Calculation of this polynomial yields the eigenvalues of A as its roots; as a matrix polynomial in the matrix A itself, it vanishes by the fundamental Cayley–Hamilton theorem. Calculating determinants, however, is computationally cumbersome, whereas this efficient algorithm is computationally significantly more efficient (in NC complexity class)."

https://en.wikipedia.org/wiki/Faddeev%E2%80%93LeVerrier_algorithm

**Comment by sritchie** [comment]
@littleredcomputer you'll like this - I realized that with the new `sicmutils.polynomial` work, I could just... return a `Polynomial` instance, which evaluates nice and fast. No `eval` needed!

**Comment by sritchie** [comment]
Closed by #463.
